source/pervasive/arithmetic/mul.rs

```diff
@@ -0,0 +1,757 @@
+/// each other when multiplied by a nonnegative integer, they have
+/// that relationship themselves
+pub proof fn lemma_mul_strict_inequality_converse_auto()
+    ensures
+        forall|x: int, y: int, z: int|
+            #[trigger]
+            (x * z) < #[trigger]
+            (y * z) && z >= 0 ==> x < y,
+{
+    assert forall|x: int, y: int, z: int|
+        #[trigger]
+        (x * z) < #[trigger]
+        (y * z) && z >= 0 implies x < y by {
+        lemma_mul_strict_inequality_converse(x, y, z);
+    }
+}
+
+/// Proof that multiplication distributes over addition, specifically that
+/// `x * (y + z) == x * y + x * z`
+pub proof fn lemma_mul_is_distributive_add(x: int, y: int, z: int)
+    ensures
+        x * (y + z) == x * y + x * z,
+{
+    MulINL::lemma_mul_is_distributive_add(x, y, z);
+}
+
+/// Proof that multiplication distributes over addition
+pub proof fn lemma_mul_is_distributive_add_auto()
+    ensures
+        forall|x: int, y: int, z: int| #[trigger] (x * (y + z)) == x * y + x * z,
+{
+    assert forall|x: int, y: int, z: int| #[trigger] (x * (y + z)) == x * y + x * z by {
+        lemma_mul_is_distributive_add(x, y, z);
+    }
+}
+
+/// Proof that multiplication distributes over addition, specifically that
+/// `(y + z) * x == y * x + z * x`
+pub proof fn lemma_mul_is_distributive_add_other_way(x: int, y: int, z: int)
+    ensures
+        (y + z) * x == y * x + z * x,
+{
+    lemma_mul_auto();
+}
+
+/// Proof that multiplication distributes over addition when the
+/// addition happens in the multiplicand (i.e., in the left-hand
+/// argument to `*`)
+proof fn lemma_mul_is_distributive_add_other_way_auto()
+    ensures
+        forall|x: int, y: int, z: int| #[trigger] ((y + z) * x) == y * x + z * x,
+{
+    assert forall|x: int, y: int, z: int| #[trigger] ((y + z) * x) == y * x + z * x by {
+        lemma_mul_is_distributive_add_other_way(x, y, z);
+    }
+}
+
+/// Proof that multiplication distributes over subtraction, specifically that
+/// `x * (y - z) == x * y - x * z`
+pub proof fn lemma_mul_is_distributive_sub(x: int, y: int, z: int)
+    ensures
+        x * (y - z) == x * y - x * z,
+{
+    lemma_mul_auto();
+}
+
+/// Proof that multiplication distributes over subtraction
+pub proof fn lemma_mul_is_distributive_sub_auto()
+    ensures
+        forall|x: int, y: int, z: int| #[trigger] (x * (y - z)) == x * y - x * z,
+{
+    assert forall|x: int, y: int, z: int| #[trigger] (x * (y - z)) == x * y - x * z by {
+        lemma_mul_is_distributive_sub(x, y, z);
+    }
+}
+
+/// Proof that multiplication distributes over subtraction when the
+/// subtraction happens in the multiplicand (i.e., in the left-hand
+/// argument to `*`). Specifically, `(y - z) * x == y * x - z * x`.
+/*
+    Note: This isn't in the Dafny standard library, but it's a logical
+    extension of [`lemma_mul_is_distributive_sub`] and
+    [`lemma_mul_is_distributive_add_other_way`].
+*/
```

Comment:
I'm not sure we necessarily need to call out places where we extend the Dafny std. lib.  Presumably we'll continue to deviate as their library and ours grow in different directions.  Maybe it would help to change the opening comment at the start of the file to something like, "It was originally based on ..."

Reply:
OK, I cut those comments.